File: domain/page/page.go
// Copyright 2016 Documize Inc. <legal@documize.com>. All rights reserved.
//
// This software (Documize Community Edition) is licensed under
// GNU AGPL v3 http://www.gnu.org/licenses/agpl-3.0.en.html
//
// You can operate outside the AGPL restrictions by purchasing
// Documize Enterprise Edition and obtaining a commercial license
// by contacting <sales@documize.com>.
//
// https://documize.com

package page

import (
	"github.com/documize/community/domain"
	"github.com/documize/community/model/page"
)

// LevelizeDocument generates level and sequence numbers for all document sections
func (h *Handler) LevelizeDocument(ctx domain.RequestContext, documentID string) {
	method := "page.LevelizeDocument"
	var err error
	ctx.Transaction.Commit()

	// Re-level all pages in document
	ctx.Transaction, err = h.Runtime.Db.Beginx()
	if err != nil {
		h.Runtime.Log.Error(method, err)
		return
	}

	p2, err := h.Store.Page.GetPages(ctx, documentID)
	if err != nil {
		ctx.Transaction.Rollback()
		h.Runtime.Log.Error(method, err)
		return
	}

	page.Levelize(p2)
	page.Sequenize(p2)

	for _, i := range p2 {
		err = h.Store.Page.UpdateLevelSequence(ctx, documentID, i.RefID, int(i.Level), i.Sequence)
		if err != nil {
			ctx.Transaction.Rollback()
			h.Runtime.Log.Error(method, err)
			return
		}
	}

	ctx.Transaction.Commit()
}

File: gui/app/models/page.js
// Copyright 2016 Documize Inc. <legal@documize.com>. All rights reserved.
//
// This software (Documize Community Edition) is licensed under
// GNU AGPL v3 http://www.gnu.org/licenses/agpl-3.0.en.html
//
// You can operate outside the AGPL restrictions by purchasing
// Documize Enterprise Edition and obtaining a commercial license
// by contacting <sales@documize.com>.
//
// https://documize.com

import { computed } from '@ember/object';
import Model from 'ember-data/model';
import attr from 'ember-data/attr';

export default Model.extend({
	documentId: attr('string'),
	orgId: attr('string'),
	contentType: attr('string'),
	pageType: attr('string'),
	level: attr('number', { defaultValue: 1 }),
	sequence: attr('number', { defaultValue: 1024 }),
	numbering: attr('string'),
	revisions: attr('number', { defaultValue: 0 }),
	blockId: attr('string'),
	title: attr('string'),
	body: attr('string'),
	rawBody: attr('string'),
	meta: attr(),
	status: attr('number', { defaultValue: 0 }),
	relativeId: attr('string'),
	userId: attr('string'),

	tagName: computed('level', function () {
		return "h2";
	}),

	tocIndent: computed('level', function () {
		return (this.get('level') - 1) * 10;
	}),

	tocIndentCss: computed('tocIndent', function () {
		let tocIndent = this.get('tocIndent');
		return `margin-left-${tocIndent}`;
	}),

	hasRevisions: computed('revisions', function () {
		return this.get('revisions') > 0;
	}),

	created: attr(),
	revised: attr(),

	// is this a new page that is pending and belongs to the user?
	isNewPageUserPending(userId) {
		return this.get('relativeId') === '' && this.get('userId') === userId && (
			this.get('status') === this.get('constants').ChangeState.PendingNew || this.get('status') === this.get('constants').ChangeState.UnderReview);
	},

	// is this new page ready for review?
	isNewPageReviewReady() {
		return this.get('relativeId') === '' && this.get('status') === this.get('constants').ChangeState.UnderReview;
	}
});
